BSP/Src/BSP_I2C.c

```diff
@@ -0,0 +1,214 @@
+    i2c_struct.Init.AddressingMode = I2C_ADDRESSINGMODE_7BIT;
+    i2c_struct.Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
+    i2c_struct.Init.GeneralCallMode = I2C_GENERALCALL_ENABLED; 
+    i2c_struct.Init.NoStretchMode = I2C_NOSTRETCH_DISABLE;
+    HAL_StatusTypeDef stat = HAL_I2C_Init(&i2c_struct);
+
+    i2c_struct.Instance = I2C1;
+    HAL_NVIC_SetPriority(I2C1_EV_IRQn, 0, 0);
+    HAL_NVIC_EnableIRQ(I2C1_EV_IRQn);
+
+    return stat;
+}
+
+/** 
+ * @brief    preform I2C write
+ * @param    hi2c               struct with config data
+ * @param    deviceAddress      target device address
+ * @param    pDataBuff          data buffer
+ * @param    len                amount of data   
+ * @return   void
+ */
+void BSP_I2C_Write(I2C_HandleTypeDef *hi2c, 
+              uint32_t deviceAddress, 
+              uint8_t* pDataBuff, 
+              uint16_t len) {
+    //copy buffer data over for security incase it changes 
+    uint8_t buffer[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH];
+        for (uint16_t i = 0; i < len; i++) {
+            buffer[i] = pDataBuff[i];
+        }
+    xSemaphoreTake(semaphore, 1);
+    if (!__HAL_I2C_GET_FLAG(hi2c, I2C_FLAG_BUSY)) {
+        // HAL requires 7b I2C address left shifted
+        HAL_I2C_Master_Transmit_IT(hi2c, deviceAddress << 1, buffer, len);
+    }
+    else if (uxQueueSpacesAvailable(I2C_DataQueue) >= len) {
+        for (uint32_t i = 0; i < len; i++) {
+            xQueueSend(I2C_DataQueue, buffer[i], 0);
+        }
+            // HAL requires 7b I2C address left shifted
+            xQueueSend(I2C_MetaQueue, deviceAddress << 1, 0);
+            xQueueSend(I2C_MetaQueue, len, 0);
+    }
+    xSemaphoreGive(semaphore);
+}
+
+/**
+ * @brief Recursively go through to see if there is any more info that needs to be transmitted,
+ * if there is transmit it
+ * @param hi2c                  I2C handle that interrupted (passed in by IRQ)
+ */
+void HAL_I2C_MasterTxCpltCallback(I2C_HandleTypeDef *hi2c) {
+    uint8_t buffer[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH];
+    if (xQueueIsQueueEmptyFromISR(I2C_MetaQueue) == pdFALSE) {
+        uint32_t deviceAddress, len;
+        uint8_t temp;
+        xQueueReceiveFromISR(I2C_MetaQueue, &deviceAddress, &temp);
+        xQueueReceiveFromISR(I2C_MetaQueue, &len, &temp);
+        for (uint32_t i = 0; i < len; i++) {
+            xQueueReceiveFromISR(I2C_DataQueue, &(buffer[i]), &temp);
+        }
+        HAL_I2C_Master_Transmit_IT(hi2c, deviceAddress, buffer, len);
+    }
+}
+
+/** 
+ * @brief    preform I2C write accounting for registers
+ * @param    hi2c               struct with config data
+ * @param    deviceAddress      target device address
+ * @param    memoryAddress      the register address to write to in the IC's memory.
+ * @param    memoryAddressSize  the register address size
+ * @param    pDataBuff          data buffer
+ * @param    len                amount of data   
+ * @return   void    
+ */
+void BSP_I2C_RegisterWrite(I2C_HandleTypeDef *hi2c, 
```

Comment:
can use success/fail to indicate queue full instead of bailing out of if/else if condition with a silent fail